Q1. Create one variable containing following type of data:
(i) string
(ii) list
(iii) float
(iv) tuple

Ans: (i) String:

In [4]:
string_variable = "PWSKILLS"
print(string_variable)

PWSKILLS


(ii) List

In [5]:
list_variable = [1, 2, 3, 4, 5]
print(list_variable)

[1, 2, 3, 4, 5]


(iii) float

In [8]:
float_variable = 3.145454
print(float_variable)

3.145454


(iv) tuple

In [12]:
tuple_variable = (1, 2, 3, 4, 5)

print(tuple_variable)

(1, 2, 3, 4, 5)


Q2. Given are some following variables containing data:
(i) var1 = ‘ ‘
(ii) var2 = ‘[ DS , ML , Python]’
(iii) var3 = [ ‘DS’ , ’ML’ , ‘Python’ ]
(iv) var4 = 1.
What will be the data type of the above given variable.

Ans: The data types of the given variables are as follows:

(i) var1 = ‘ ‘ => The data type of var1 is a string. It contains a single space character.

(ii) var2 = ‘[ DS , ML , Python]’ => The data type of var2 is also a string. It contains
    a sequence of characters enclosed in single quotes.

(iii) var3 = [ ‘DS’ , ’ML’ , ‘Python’ ] => The data type of var3 is a list. It contains
     multiple elements enclosed in square brackets. Each element in the list is a string.

(iv) var4 = 1. => The data type of var4 is a float. The presence of a decimal point after
    the number 1 indicates that it is a floating-point number.

Q3. Explain the use of the following operators using an example:
(i) /
(ii) %
(iii) //
(iv) **

Ans: The explanation of each operator along with an example are as under: 


(i) / (Division Operator) : The division operator / performs division between two operands 
    and returns the quotient as a floating-point number. It is used to divide one number by another.

Example:

In [19]:
a = 112
b = 22
result = a / b
print(result)

5.090909090909091


In this example, the division operator / divides a by b, resulting in the quotient 5.090909090909091.

(ii) % (Modulo Operator): The modulo operator % returns the remainder after the division of 
    one number by another. It is used to obtain the remainder of a division operation.

Example:

In [20]:
a = 10
b = 3
result = a % b
print(result)


1


In this example, the modulo operator % calculates the remainder when a is divided by b,
resulting in the value 1.

(iii) // (Floor Division Operator): The floor division operator // performs division between 
     two operands and returns the quotient as an integer, rounded down to the nearest whole 
     number. It discards the fractional part of the division result.

Example:

In [21]:
a = 10
b = 3
result = a // b
print(result)

3


In this example, the floor division operator // divides a by b and returns the whole number quotient 3,
disregarding the fractional part

(iv) ** (Exponentiation Operator): The exponentiation operator ** raises a base number to the power of 
    an exponent. It calculates the result of exponential operations.

Example:

In [22]:
base = 2
exponent = 3
result = base ** exponent
print(result)

8


In this example, the exponentiation operator ** raises base to the power of exponent, 
resulting in 8.

These operators (/, %, //, **) are fundamental mathematical operators used in Python for various calculations
and manipulations involving numbers.

Q4. Create a list of length 10 of your choice containing multiple types of data. Using for loop print the
element and its data type.

Ans: Creating a list of length 10 with different types of data and using a for loop to print each
     element along with its data type:

In [23]:
my_list = [10, 'Hello', 3.14, True, [1, 2, 3], {'name': 'John'}, (4, 5, 6), None, 'pwskills', 2023]

for element in my_list:
    print(element, type(element))


10 <class 'int'>
Hello <class 'str'>
3.14 <class 'float'>
True <class 'bool'>
[1, 2, 3] <class 'list'>
{'name': 'John'} <class 'dict'>
(4, 5, 6) <class 'tuple'>
None <class 'NoneType'>
pwskills <class 'str'>
2023 <class 'int'>


Q5. Using a while loop, verify if the number A is purely divisible by number B and if so then how many
times it can be divisible.

Ans: Using a while loop to verify if number A is divisible by number B and counting how many times it can be divided:

In [24]:
A = 36
B = 4
count = 0

while A % B == 0:
    A = A // B
    count += 1

print("Number A can be divided by B", count, "times.")

Number A can be divided by B 1 times.


Q6. Create a list containing 25 int type data. Using for loop and if-else condition print if the element is
divisible by 3 or not.

Ans: Create a list containing 25 int type data. Using for loop and if-else condition print if the element is
divisible by 3 or not :-

In [25]:
my_list = [7, 12, 21, 8, 33, 18, 42, 10, 27, 15, 5, 30, 9, 16, 25, 36, 6, 13, 39, 22, 4, 31, 20, 24, 11]

for element in my_list:
    if element % 3 == 0:
        print(element, "is divisible by 3")
    else:
        print(element, "is not divisible by 3")


7 is not divisible by 3
12 is divisible by 3
21 is divisible by 3
8 is not divisible by 3
33 is divisible by 3
18 is divisible by 3
42 is divisible by 3
10 is not divisible by 3
27 is divisible by 3
15 is divisible by 3
5 is not divisible by 3
30 is divisible by 3
9 is divisible by 3
16 is not divisible by 3
25 is not divisible by 3
36 is divisible by 3
6 is divisible by 3
13 is not divisible by 3
39 is divisible by 3
22 is not divisible by 3
4 is not divisible by 3
31 is not divisible by 3
20 is not divisible by 3
24 is divisible by 3
11 is not divisible by 3


Q7. What do you understand about mutable and immutable data types? Give examples for both showing
this property.

Ans: * Immutable Data Types: Immutable data types are those whose state cannot be modified after they are created. Any operation on an immutable object results in the creation of a new object rather than modifying the original one. Examples of immutable data types include strings, numbers (integers, floats), and tuples.

In [26]:
# Example of immutable data types

# Strings
name = "Alice"
new_name = name.replace("A", "E")
print(name)  # Output: "Alice"
print(new_name)  # Output: "Elise"

# Numbers
num = 10
num += 5
print(num)  # Output: 15

# Tuples
my_tuple = (1, 2, 3)
new_tuple = my_tuple + (4, 5)
print(my_tuple)  # Output: (1, 2, 3)
print(new_tuple)  # Output: (1, 2, 3, 4, 5)

Alice
Elice
15
(1, 2, 3)
(1, 2, 3, 4, 5)


* Mutable Data Types: Mutable data types are those that can be modified after they are created. 
 Operations performed on mutable objects directly modify the object itself without creating a new
object. Examples of mutable data types include lists, dictionaries, and sets.

In [27]:
# Example of mutable data types

# Lists
my_list = [1, 2, 3]
my_list.append(4)
print(my_list)  # Output: [1, 2, 3, 4]

# Dictionaries
my_dict = {'name': 'Alice', 'age': 25}
my_dict['age'] = 26
print(my_dict)  # Output: {'name': 'Alice', 'age': 26}

# Sets
my_set = {1, 2, 3}
my_set.add(4)
print(my_set)  # Output: {1, 2, 3, 4}

[1, 2, 3, 4]
{'name': 'Alice', 'age': 26}
{1, 2, 3, 4}
